bokehjs/src/lib/models/glyphs/vsteparea.ts

```diff
@@ -0,0 +1,184 @@
+import {Area, AreaView, AreaData} from "./area"
+import {Context2d} from "core/util/canvas"
+import {SpatialIndex} from "core/util/spatial"
+import * as hittest from "core/hittest"
+import * as p from "core/properties"
+import {StepMode} from "core/enums"
+import {unreachable} from "core/util/assert"
+import {Selection} from "../selections/selection"
+
+export type VStepAreaData = AreaData & {
+  _x: FloatArray
+  _y1: FloatArray
+  _y2: FloatArray
+
+  sx: ScreenArray
+  sy1: ScreenArray
+  sy2: ScreenArray
+}
+
+export interface VStepAreaView extends VStepAreaData {}
+
+export class VStepAreaView extends AreaView {
+  declare model: VStepArea
+  declare visuals: VStepArea.Visuals
+
+  protected _index_data(index: SpatialIndex): void {
+    const {min, max} = Math
+
+    for (let i = 0; i < this.data_size; i++) {
+      const x = this._x[i]
+      const y1 = this._y1[i]
+      const y2 = this._y2[i]
+      index.add_rect(x, min(y1, y2), x, max(y1, y2))
+    }
+
+  }
+
+  protected _step_path(ctx: Context2d, mode: StepMode, sx: ScreenArray, sy: ScreenArray, from_i: number, to_i: number): void {
+    // Assume that path was already moved to the first point
+    let prev_x = sx[from_i]
+    let prev_y = sy[from_i]
+    const idx_dir = from_i < to_i ? 1 : -1
+    for (let i = from_i + idx_dir; i != to_i; i += idx_dir) {
+      switch (mode) {
+        case "before":
+          ctx.lineTo(prev_x, sy[i])
+          ctx.lineTo(sx[i], sy[i])
+          break
+        case "after":
+          ctx.lineTo(sx[i], prev_y)
+          ctx.lineTo(sx[i], sy[i])
+          break
+        case "center":
+          const mid_x = (prev_x + sx[i]) / 2
+          ctx.lineTo(mid_x, prev_y)
+          ctx.lineTo(mid_x, sy[i])
+          ctx.lineTo(sx[i], sy[i])
+          break
+        default:
+          unreachable()
+      }
+      prev_x = sx[i]
+      prev_y = sy[i]
+    }
+  }
+
+  protected _flip_mode(mode: StepMode): StepMode {
+    switch (mode) {
+      case "before":
+        return "after"
+      case "after":
+        return "before"
+      case "center":
+        return "center"
+      default:
+        unreachable()
+    }
+  }
+
+  protected _render(ctx: Context2d, _indices: number[], data?: VStepAreaData): void {
+    const {sx, sy1, sy2} = data ?? this
+
+    const forward_mode = this.model.step_mode
+    const backward_mode = this._flip_mode(this.model.step_mode)
+
+    ctx.beginPath()
+    ctx.moveTo(sx[0], sy1[0])
+
+    this._step_path(ctx, forward_mode, sx, sy1, 0, sx.length)
+    this._step_path(ctx, backward_mode, sx, sy2, sx.length, -1)
+
+    ctx.closePath()
+
+    this.visuals.fill.apply(ctx)
+    this.visuals.hatch.apply(ctx)
+  }
+
+  scenterxy(i: number): [number, number] {
+    const scx = this.sx[i]
+    const scy = (this.sy1[i] + this.sy2[i])/2
+    return [scx, scy]
+  }
+
+  protected override _hit_point(geometry: PointGeometry): Selection {
+    const {sx, sy1, sy2} = this
+
+    for (let i = 0; i < this.sx.length - 1; i++) {
```

Comment:
It more idiomatic to use `this.data_size` instead of `this.sx.length`, similarly to how it is used in `_index_data()`.

Reply:
Changed. :+1: